releases_page/main.go

```diff
@@ -0,0 +1,164 @@
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"html/template"
+	"io"
+	"net/http"
+	"os"
+)
+
+// Templates for main page and release page
+const (
+	mainPageTemplate = `<ul>
+	<li>
+		<a href="../">../</a>
+	</li>
+	{{ range . }}
+	<li>{{ $releaseName := .name | trimV }}
+		<a href="./{{ $releaseName }}/">{{ printf "tofu_%s" $releaseName }}</a>
+	</li>
+	{{ end }}
+	</ul>`
+	releasePageTemplate = `<ul>
+	<li>
+		<a href="../">../</a>
+	</li>
+	{{ range . }}
+	<li>
+		<a href="{{ .browser_download_url }}">{{ .name }}</a>
+	</li>
+	{{end}}
+	</ul>`
+)
+
+func main() {
+	os.Exit(realMain())
+}
+
+func realMain() int {
+	const (
+		githubAPIEndpoint = "https://api.github.com/repos/opentofu/opentofu/releases"
+		htmlFileName      = "index.html"
+	)
+
+	// Check for GitHub token in environment variable
+	token := os.Getenv("GITHUB_TOKEN")
+	if token == "" {
+		fmt.Println("Failed to retrieve Authorization token from environment variable. Please ensure that 'GITHUB_TOKEN' is set properly.")
+		return 1
+	}
+
+	// Create client
+	client := &http.Client{}
+	req, err := http.NewRequest("GET", githubAPIEndpoint, nil)
+	if err != nil {
+		fmt.Println("Error creating request: ", err)
+		return 1
+	}
+	// Set custom user-agent in request header
+	req.Header.Set("User-Agent", "opentofu releases page")
+	// Set token in request header
+	req.Header.Set("Authorization", "token "+token)
+
+	// Fetch releases from GitHub API
+	response, err := client.Do(req)
+	if err != nil {
+		fmt.Println("Error fetching releases: ", err)
+		return 1
+	}
+	defer response.Body.Close()
+
+	body, err := io.ReadAll(response.Body)
+	if err != nil {
+		fmt.Println("Error reading response body: ", err)
+		return 1
+	}
+
+	// Unmarshal JSON response
+	releases := []map[string]interface{}{}
+	if err := json.Unmarshal(body, &releases); err != nil {
+		fmt.Println("Error unmarshalling JSON: ", err)
+		return 1
+	}
+
+	// Create main index.html file
+	htmlFile, err := os.Create(htmlFileName)
+	if err != nil {
+		fmt.Println("Error creating main HTML file: ", err)
+		return 1
+	}
+
+	// Function to remove the first character of a string
+	funcMap := template.FuncMap{
+		"trimV": func(s string) string {
+			if len(s) <= 1 {
+				return ""
+			}
+			return s[1:]
+		},
+	}
+
+	// Parse mainPage template
+	mainPageTmpl, err := template.New("mainPageTemplate").Funcs(funcMap).Parse(mainPageTemplate)
+	if err != nil {
+		fmt.Println("Error parsing mainPage template: ", err)
```

Comment:
This branch does not close the file properly.

Reply:
This code snippet is not using any file, I am not sure about which file you are talking about.

Reply:
You open the `htmlFile` above, but it is not closed before the return.

Reply:
will do that